### PR TITLE
Check the args when init class and fix some typo.

### PR DESCRIPTION
**Pull request by ponytailer**
- throw the errors earlier
- ~common line_length = 79 in `black config`~

### CODE REVIEW

**Comment by abersheeran** [comment]
To be honest, I don't like the 79 limit, so I didn't configure this option before.

File: ratelimit/core.py [hunk 1]
```diff
@@ -34,50 +36,69 @@ def __init__(
...
+
+        assert isinstance(
+            backend, BaseBackend
+        ), f"invalid backend: {self.backend}"
+
+        try:
+            self.config: Dict[re.Pattern, Sequence[Rule]] = {
+                re.compile(path): value for path, value in config.items()
+            }
+        except re.error as exc:
```
Comment (abersheeran) [inline ratelimit/core.py:61]: The capture here is redundant, I think `re.error` itself provides clear enough error information.

File: ratelimit/core.py [hunk 2]
```diff
@@ -34,50 +36,69 @@ def __init__(
...
+
         self.app = app
         self.authenticate = authenticate
         self.backend = backend
-        self.config: Dict[re.Pattern, Sequence[Rule]] = {
-            re.compile(path): value for path, value in config.items()
-        }
+
+        assert isinstance(
+            self.authenticate, Callable
```
Comment (abersheeran) [inline ratelimit/core.py:50]: Great idea, but it's better to change to `asyncio.iscoroutinefunction(......)`.

**Comment by ponytailer** [comment]
> To be honest, I don't like the 79 limit, so I didn't configure this option before.

ok, I remove it.

File: ratelimit/core.py [hunk 3]
```diff
@@ -52,32 +61,31 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+            # Select the first rule that can be matched
+            match_rule = list(filter(lambda r: r.group == group, rules))
+            if match_rule:
+                rule = match_rule[0]
+                break
         else:  # If no rule can match, run `self.app` and return
             return await self.app(scope, receive, send)
 
-        if not [name for name in RULENAMES if getattr(rule, name) is not None]:
+        if not list(filter(lambda name: getattr(rule, name) is not None, RULENAMES)):
```
Comment (abersheeran) [inline ratelimit/core.py:84]: What is the purpose of changing the comprehension into a function call? It does not make the code faster.

Comment (abersheeran) [reply ratelimit/core.py:84]: Inspired by another PR, it should be better to use `any` here.

Comment (ponytailer) [reply ratelimit/core.py]: Ok, I will check it.

**Comment by ponytailer** [comment]
BTW,  can support simple rate-limit in memory?

**Comment by abersheeran** [comment]
Yes, we can add a new memory backend. In fact, I wanted to add this function at first, but I've been busy with other work. If you are willing to submit PR, I will be happy to help you complete it together.

**Comment by ponytailer** [comment]
> Yes, we can add a new memory backend. In fact, I wanted to add this function at first, but I've been busy with other work. If you are willing to submit PR, I will be happy to help you complete it together.

OK, I will add it later.

**Pull request merged**